clients/pkg/promtail/wal/reader.go

```diff
@@ -0,0 +1,61 @@
+package wal
+
+import (
+	"fmt"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+
+	"github.com/grafana/loki/pkg/ingester"
+	"github.com/grafana/loki/pkg/util"
+	wal2 "github.com/grafana/loki/pkg/util/wal"
```

Comment:
unfortunately this code is AGPL so we'll need to either need to go through the necessary bits to move it to a different path that can be licensed differently, or duplicate the functionality

Reply:
Re-licensed this bit as discussed. This [PR](https://github.com/grafana/loki/pull/8315) contains those changes to keep them separate. Will rebase after.